### DISCUSSION TITLE
Use GET method instead of POST during submit button

### DISCUSSION THREAD

**Discussion by carlogravador** [jrief/django-formset#80] [Q&A]
Is it possible to invoked GET method instead of POST when submitting a django-formset form.

I'm creating a filter using django-filterset and rendering the said form in my template using `render_form` tag.

But everytime I click the submit button to filter my queryset, I can see that POST method is getting invoked.

template.html
```html
<div class="form-group">
    <div class="card">
        <h3>Filter By:</h3>
        <br>
        <!--TODO: Fix style of filter forms-->
        <django-formset endpoint="{{ request.path }}" csrf-token="{{ csrf_token }}">
        {% render_form filterset.form "bootstrap" field_classes="pb-3 mb-2" %}

        <div class="d-flex align-items-center justify-content-md-end">
            <button class="primaryAction btn btn-primary" click="disable -> spinner -> submit -> proceed !~ scrollToError">Submit</button>
        </div>

        </django-formset>
    </div>
</div>

```

**Comment by jrief** [comment]
**django-formset** always transfers the form(s) payload using encoding type `application/json` hence using method `GET` doesn't make any sense, because then the URL would become long, unreadable and full of escape characters.

**Reply by Actionb** [reply]
@jrief  Does an URL have to be readable though? It is plausible that if you use a GET form that you **want** the form data to be encoded in the URL. For example when using a filter or search form (like it is the case here), because it lets you bookmark the results.

Is it possible for a django-formset form to just submit the form data as a GET request without first sending the data as a POST request?  
  I tinkered around a little bit, but couldn't find a way. If I wanted to keep using the django-formset form renderers, I also had to use the django-formset submit buttons, which means I couldn't use a plain GET request to submit the form. In the end, I'd have to use a plain django form view - and if I wanted to have the form rendered with a CSS framework, I'd have to use a different library or write my own renderer. Seems a bit silly when all I want is to have the form be submitted with the request type specified in the form.

**Reply by Actionb** [reply]
**How would you implement a search form in django-formset?**

Assume a search form is a form where the input will be used to filter a the queryset of a list view.  The form method is GET, because you merely request resources from the server - you do not create or update any.
 
Submitting such a search form in django-formset is a two-step process where the view is called two times: `submit (POST) -> proceed (GET)`. `submit` gets the search form data and is meant to validate it. `proceed` is meant to return the page with the results. But proceed does not know the search parameters; proceed is just a request against the success_url from the first step. 

To let proceed know how to filter, you either have to store the search form data from the submit step in the session or you have to include the data in the success_url in the form of a query string. The latter is pretty much exactly the same thing as just submitting the form with a GET request in the first place. And the former does not produce an URL with a query string of the form data, which means that the user can't easily bookmark the search results (they'd have to write the query string themselves I guess?).

**Comment by carlogravador** [comment]
Is it possible then to redirect the data from POST method to GET method, as for filtering a certain queryset, GET is the method that is used?

**Comment by SamuelJennings** [comment]
Was just having a go at implementing this myself because I need the same functionality for a filter view like OP. If you're willing to forgo the use of any df-click actions on the submit button, then you can submit as a GET request using some simple inline javascript.

The javascript is required to **force** form submission because the regular 'submit' event is being prevented by django-formset.

So here's my solution for a GET request including URL params:

    <django-formset endpoint="{{ request.path }}" csrf-token="{{ csrf_token }}">
        {% render_form filterset.form %}
        <button type="submit" form="{{ filterset.form.form_id }}" onclick="document.getElementById('{{filterset.form.form_id}}').submit();">Submit</button>
    </django-formset>

Not sure if there are any side effects but seems to work fine for me :)

**Comment by jrief** [comment]
@SSJenny90 Your solution is, hmm, a bit hacky. However, I get your point. May I ask why in the first place you want to use django-formset to create forms invoked using method="GET"? If it's because you want to have it rendered using the Bootstrap renderer, then I can look for a solution which works for you.

**Reply by SamuelJennings** [reply]
> Your solution is, hmm, a bit hacky. 

Haha sure but at least it's a relatively simple one!

> If it's because you want to have it rendered using the Bootstrap renderer, then I can look for a solution which works for you.

I mean yeah, being able to use a consistent renderer across the entire site would be a huge plus I think. Especially for 3rd party forms that don't account for django-formset in their views. Obviously one would sacrifice some functionality in this case but at least everything would look consistent.

> May I ask why in the first place you want to use django-formset to create forms invoked using method="GET"? 

Well to start, I rely on django-filter for filtering list views and it expects a GET request. I could modify the FilterView class to work on post but the javascript hack above was much easier. 

Also, like @Actionb said, method="GET" allows one to bookmark (or share) a URL inclusive of query params. I build web portals for research databases so the ability for my users to share advanced searches with their colleagues via a simple link (or FB, Twitter, LinkedIn, etc.) would be a huge benefit.

**Reply by jrief** [reply]
Okay, then the best option will be to adopt the current renderers (which one of then do you actually use?) to render a `<form method="GET">` and send the form fields through standard query params. Encoding JSON inside a GET request just doesn't feel right for me.

**Reply by SamuelJennings** [reply]
Hmmm you know what, I might have over-engineered my last solution. The only requirement is to direct the submit button towards the correct form using the form attribute because the button will always occur outside the form element tags due how everything is rendered. 

    <django-formset endpoint="{{ request.path }}" csrf-token="{{ csrf_token }}">
        {% render_form filter.form %}
        <button type="submit" form="{{filter.form.form_id}}">Apply Filters</button>
    </django-formset>

This is pretty standard stuff IMO and therefore I'd say it's not necessary to start modifying renderers to account for this use case. Thoughts?

Not sure why but I was convinced that something in the included JS module was preventing the standard submit event for forms inside the django-formset tags. Doesn't seem to be the case though.

**Comment by Alex-Sichkar** [comment]
+1 to this! Using [django-filter](https://django-filter.readthedocs.io/en/stable/index.html) has become the de facto standard for integrating filtering into Django projects. An additional advantage of using a `django-formset` form as a FilterSet form is the ability to reuse widgets—most notably, `Selectize`. Otherwise, you'd need to integrate an alternative autocomplete solution, which would require extra work for styling and configuration to achieve a consistent look and feel. Additionally, maintenance would be more challenging.

**Reply by jrief** [reply]
I will have a look at this. Adding support for django-filter is a strong argument I agree upon. Will not be available before version 1.8, so please be patient.

**Reply by jrief** [reply]
Version 1.7.4 has support for **django-filter**. Please test this in your environment.

**Reply by Alex-Sichkar** [reply]
Great, thanks! I'm going to test it soon!

**Reply by Alex-Sichkar** [reply]
I have tried it, and it seems to work just fine. However, when I mentioned `django-filter` integration, I was referring to filtering on list views. I have no idea how it should work and am not sure if it's possible because `django-formset` uses POST + JSON. To filter lists, GET is required to capture the filter URL.

Anyway, thank you!
